### ISSUE TITLE
How to expose to Excel functions from a package?

### ISSUE THREAD

**Issue by mnar53** [sdllc/Basic-Excel-R-Toolkit#12]
I'm absolutely a newbie to R, so forgive if I'm asking a stupid question: is there a way to expose directly all functions from a package? For example: I'm trying to expose to Excel special functions from package GSL; I could manually write interface function in Functions.R for them, but they are literally  hundreths.
I don't know R sufficiently to make this automatic/programmatic.

P.S.: in this particular case, I've found that in the package source there are individual files (e.g., "bessel.R"): functions are defined there, and they simply call the gsl dll. So I could just source these files in Functions.R, and probably this would expose bessel functions to Excel. But this seems a kind of hack, and probably cannot be always applid.

**Comment by duncanwerner** [comment]
There's no automatic way to do this, and there are reasons not to do it.  But if you wanted to do it, one way to do it would be something like

```R
library(gsl);
lapply( ls( as.environment( 'package:gsl' )), function(name){
	newname <- paste0( "GSL.", name );
	cat(paste0(newname, "\n"));
	func <- get( name, envir=as.environment( 'package:gsl' ));
	if( is.function( func )){
		assign( newname, func, .GlobalEnv );
	}
});
```

You should probably rename the functions (as I did here) but it's not strictly required.  

Now why not do this?

(1) We don't support enough functions, so many of them will disappear.  This is because we have an arbitrary limit on the functions, as discussed in #8.  We will probably increase this in a future release (or you could do it).

(2) Although it's not the case here, often times R functions will use parameters that we can't support in Excel, like functions.  So those just won't work.  They won't break, though (I don't think so, anyway -- should be tested).

Still it's probably useful in some cases.  As discussed in #9 it would be useful to put these in a separate Excel category.  If we add function space, I think we'll add a method for exposing an environment in one go, which could be used here.

TL;DR: use the function above, but there aren't enough function slots (for now).

**Comment by mnar53** [comment]
By the way, today the best homework I could do was something like:

```
LISTPKG <- function (pkgname) {
  x <- ls(pos=paste0("package:",pkgname))
  paste(pkgname, '.', x, " <- ", pkgname, '::', x)
}
```
I call it from Excel, and then paste the result back in Functions.R.
Your solution, till Excel categories aren't .available, is more streamlined and satisfatory.

**Comment by mnar53** [comment]
Hi Duncan,

thank you for your kind reply. Just some points about my current setup:

1. I've already increased the number of slots to 2,000. After all, only changes required are:

Increase of MAX_FUNCTION_COUNT to 2000, line 72 of BERT.hAdd lines to BERT.def, till BERTFunctionCall2999Add line to BERT.cpp, till BFC(2999);

I've compiled it with Intel 14.0, release x64, eventually compressed the dll with upx, and it works fine.


2. I'm currently using Microsoft Revolution R, which is multithreaded and fast.  That is:


I've uninstalled any previous version of RI've installed BERTand then, installed MRO under BERT\R-3.2.3

at least, I've just one single copy of R into which to install packages.


3. In Functions.R among others I've the following code:


usepkg <- function (pkg) {
    library(pkg,character.only=TRUE)
    env = as.environment(paste0('package:',pkg ))
    lapply( ls(env,pkg ), function(name){
        newname <- paste0( pkg, '.', name );
        cat(paste0(newname, "\n"));
        func <- get( name, envir=env);
        if( is.function( func )){
            assign( newname, func, .GlobalEnv );
        }
    });
}    

usepkg("gsl")
usepkg("Matrix")
usepkg("pracma")
usepkg("matrixcalc")
usepkg("gaussquad")
usepkg("signal")
usepkg("matlab")
usepkg("tseries")
usepkg("forecast")


This means A LOT of functions available to Excel, but even so the startup is fast and quite responsive.


4. Even if banal, let me point how to navigate through all these functions:


In a cell, I type =R.signal, in the popup I choose R.signal.ArmaI press the Excel Fx button, and I get the poupbox with the function arguments to give!



Nothing of special about all this, but I think it helps to make the R-Excel toolkit good-to-use,
and competitive (or even better) than other solutions, like Excel-Python (with link to numpy/scipy).


Thanks,
   maurizio



















Date: Fri, 15 Apr 2016 06:47:19 -0700

**Comment by duncanwerner** [comment]
We now support this directly, via two methods `BERT$UseEnvironment` and `BERT$UsePackage`.  The only reason to use these rather than the function you describe above is because they reference the functions within their environment, instead of copying.

`BERT$UseEnvironment( env, prefix, ... )`
`BERT$UsePackage( env, prefix, ... )`

`UsePackage` is an alias that prepends `pacakge:` to the environment name.  The `...` parameters are passed to `ls()`, so you can map a subset of functions if you want, e..g

`BERT$UsePackage("gsl", "GSL", pattern="bessel");`

which will map only the Bessel functions.
